docs/developer-guide/nodejs-api.md

```diff
@@ -240,7 +240,15 @@ The return value is an object containing the results of the linting operation. H
 }
 ```
 
-The top-level report object has a `results` array containing all linting results for files that had warnings or errors (any files that did not produce a warning or error are omitted). Each file result includes the `filePath`, a `messages` array, `errorCount`, `warningCount`, and optionally `output`. The `messages` array contains the result of calling `linter.verify()` on the given file. The `errorCount` and `warningCount` give the exact number of errors and warnings respectively on the given file. The `output` property gives the source code for the file with as many fixes applied as possible, so you can use that to rewrite the files if necessary. The top-level report object also has `errorCount` and `warningCount` which give the exact number of errors and warnings respectively on all the files.
+The top-level report object has a `results` array containing all linting results for files that had warnings or errors (any files that did not produce a warning or error are omitted). Each file result includes:
+
+* `filePath` - Path to the given file.
+* `messages` - Array containing the result of calling `linter.verify()` on the given file.
+* `errorCount` and `warningCount` - The exact number of errors and warnings respectively on the given file.
+* `source` - The source code for the given file. This property is omitted if this file has no errors/warnings or if the `output` property is present.
```

Comment:
Sorry, I should have explained this as well from the get-go. Apologies for not being clear enough.

In my initial implementation, `source` was indeed the text before fixes.
But all of the error/warning messages in the `messages` property contain the line/column location **after** the fixes, so if we would use the `source` property in that case, the formatter would potentially report an incorrect location for each error.

As an example, let's say there are two messages, the first can be fixed, the second one cannot:
```javascript
[{
    message: 'Expected line before comment',
    line: 4,
    column: 1
},
{
    message: '\'foo\' is defined but never used.',
    line: 5,
    column: 7
}]
```

After the first message is fixed, a newline is added above line 4 in the source code. The messages array is then updated to contain only one message:

```javascript
[{
    message: '\'foo\' is defined but never used.',
    line: 6, // <--- line is updated to reflect fixes
    column: 7
}]
```

-- 

This means that the formatter would have to check for the presence of `output` in that case anyway, even if the `source` property would not be omitted when there are fixes.